### PR TITLE
refactor: migrate from `from_proto_bytes` to `value_encoding`

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- migrate from `from_proto_bytes` to `value_encoding` while encoding to/decoding from storage
- fix the bug that default values in struct and list are converted to NULLs caused by  `from_proto_bytes`
- in addition, serialization of value encoding for List and Struct are faster than key encoding now because the overhead of protobuf encoding is eliminated, but deserialization is still slower(though much faster than before).

## Checklist

~- [ ] I have written necessary rustdoc comments~
~- [ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#4521 
#4672 
#5713

### CODE REVIEW

**Comment by waruto210** [comment]
benchmark result on my computer:
```shell
Int16 key encoded len: 3
bench Int16 (key encoding serialization)
                        time:   [15.289 ns 15.354 ns 15.422 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Int16 value encoded len: 3
bench Int16 (value encoding serialization)
                        time:   [14.164 ns 14.223 ns 14.290 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Int32 key encoded len: 5
bench Int32 (key encoding serialization)
                        time:   [15.226 ns 15.288 ns 15.353 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Int32 value encoded len: 5
bench Int32 (value encoding serialization)
                        time:   [13.324 ns 13.379 ns 13.459 ns]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

Int64 key encoded len: 9
bench Int64 (key encoding serialization)
                        time:   [30.577 ns 30.728 ns 30.890 ns]

Int64 value encoded len: 9
bench Int64 (value encoding serialization)
                        time:   [24.752 ns 24.810 ns 24.867 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Float32 key encoded len: 5
bench Float32 (key encoding serialization)
                        time:   [15.217 ns 15.308 ns 15.407 ns]

Float32 value encoded len: 5
bench Float32 (value encoding serialization)
                        time:   [13.518 ns 13.553 ns 13.589 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

Float64 key encoded len: 9
bench Float64 (key encoding serialization)
                        time:   [29.953 ns 30.018 ns 30.087 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild

Float64 value encoded len: 9
bench Float64 (value encoding serialization)
                        time:   [25.081 ns 25.179 ns 25.273 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Bool key encoded len: 2
bench Bool (key encoding serialization)
                        time:   [14.753 ns 14.820 ns 14.886 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

Bool value encoded len: 2
bench Bool (value encoding serialization)
                        time:   [13.564 ns 13.605 ns 13.648 ns]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

Decimal key encoded len: 5
bench Decimal (key encoding serialization)
                        time:   [61.071 ns 61.218 ns 61.369 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

Decimal value encoded len: 17
bench Decimal (value encoding serialization)
                        time:   [28.035 ns 28.167 ns 28.338 ns]
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

Interval key encoded len: 17
bench Interval (key encoding serialization)
                        time:   [59.241 ns 59.401 ns 59.563 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild

Interval value encoded len: 17
bench Interval (value encoding serialization)
                        time:   [56.623 ns 56.750 ns 56.878 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

NaiveDate key encoded len: 5
bench NaiveDate (key encoding serialization)
                        time:   [14.716 ns 14.752 ns 14.787 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe

NaiveDate value encoded len: 5
bench NaiveDate (value encoding serialization)
                        time:   [13.767 ns 13.812 ns 13.859 ns]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

NaiveDateTime key encoded len: 13
bench NaiveDateTime (key encoding serialization)
                        time:   [30.853 ns 30.967 ns 31.088 ns]
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) low mild
  6 (6.00%) high mild
  1 (1.00%) high severe

NaiveDateTime value encoded len: 13
bench NaiveDateTime (value encoding serialization)
                        time:   [26.167 ns 26.237 ns 26.318 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

NaiveTime key encoded len: 9
bench NaiveTime (key encoding serialization)
                        time:   [30.295 ns 30.364 ns 30.439 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

NaiveTime value encoded len: 9
bench NaiveTime (value encoding serialization)
                        time:   [24.984 ns 25.037 ns 25.091 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

Utf8 (len = 10) key encoded len: 20
bench Utf8 (len = 10) (key encoding serialization)
                        time:   [66.011 ns 66.224 ns 66.446 ns]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

Utf8 (len = 10) value encoded len: 15
bench Utf8 (len = 10) (value encoding serialization)
                        time:   [26.115 ns 26.198 ns 26.284 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Utf8 (len = 1000) key encoded len: 1127
bench Utf8 (len = 1000) (key encoding serialization)
                        time:   [850.89 ns 852.82 ns 854.79 ns]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

Utf8 (len = 1000) value encoded len: 1005
bench Utf8 (len = 1000) (value encoding serialization)
                        time:   [66.217 ns 66.396 ns 66.582 ns]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild

Utf8 (len = 10000) key encoded len: 112502
bench Utf8 (len = 10000) (key encoding serialization)
                        time:   [61.335 µs 61.529 µs 61.736 µs]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

Utf8 (len = 10000) value encoded len: 100005
bench Utf8 (len = 10000) (value encoding serialization)
                        time:   [1.6608 µs 1.6653 µs 1.6700 µs]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe

Struct of Bool (len = 100) key encoded len: 201
bench Struct of Bool (len = 100) (key encoding serialization)
                        time:   [1.3220 µs 1.3248 µs 1.3276 µs]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

Struct of Bool (len = 100) value encoded len: 201
bench Struct of Bool (len = 100) (value encoding serialization)
                        time:   [725.14 ns 726.89 ns 728.74 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

List of Bool (len = 100) key encoded len: 227
bench List of Bool (len = 100) (key encoding serialization)
                        time:   [1.6736 µs 1.6771 µs 1.6808 µs]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

List of Bool (len = 100) value encoded len: 205
bench List of Bool (len = 100) (value encoding serialization)
                        time:   [699.31 ns 700.93 ns 702.68 ns]
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

bench Int16 (key encoding deserialization)
                        time:   [4.3788 ns 4.4010 ns 4.4281 ns]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

bench Int16 (value encoding deserialization)
                        time:   [3.3505 ns 3.3591 ns 3.3684 ns]
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

bench Int32 (key encoding deserialization)
                        time:   [4.4010 ns 4.4140 ns 4.4278 ns]
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  9 (9.00%) high mild
  2 (2.00%) high severe

bench Int32 (value encoding deserialization)
                        time:   [3.3382 ns 3.3460 ns 3.3537 ns]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

bench Int64 (key encoding deserialization)
                        time:   [4.5358 ns 4.5506 ns 4.5664 ns]
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

bench Int64 (value encoding deserialization)
                        time:   [3.3930 ns 3.4050 ns 3.4186 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

bench Float32 (key encoding deserialization)
                        time:   [4.3750 ns 4.3831 ns 4.3910 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

bench Float32 (value encoding deserialization)
                        time:   [3.3390 ns 3.3453 ns 3.3521 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

bench Float64 (key encoding deserialization)
                        time:   [4.5830 ns 4.5911 ns 4.5992 ns]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

bench Float64 (value encoding deserialization)
                        time:   [3.3331 ns 3.3391 ns 3.3453 ns]
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe

bench Bool (key encoding deserialization)
                        time:   [4.1757 ns 4.1851 ns 4.1959 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

bench Bool (value encoding deserialization)
                        time:   [3.3413 ns 3.3504 ns 3.3606 ns]
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

bench Decimal (key encoding deserialization)
                        time:   [37.286 ns 37.358 ns 37.430 ns]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

bench Decimal (value encoding deserialization)
                        time:   [7.0831 ns 7.0988 ns 7.1149 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

bench Interval (key encoding deserialization)
                        time:   [5.4229 ns 5.4347 ns 5.4462 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

bench Interval (value encoding deserialization)
                        time:   [3.9637 ns 3.9721 ns 3.9810 ns]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

bench NaiveDate (key encoding deserialization)
                        time:   [7.0366 ns 7.0492 ns 7.0621 ns]
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe

bench NaiveDate (value encoding deserialization)
                        time:   [5.8505 ns 5.8650 ns 5.8808 ns]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

bench NaiveDateTime (key encoding deserialization)
                        time:   [9.6108 ns 9.6347 ns 9.6606 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

bench NaiveDateTime (value encoding deserialization)
                        time:   [7.7553 ns 7.7738 ns 7.7921 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

bench NaiveTime (key encoding deserialization)
                        time:   [4.8020 ns 4.8148 ns 4.8286 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

bench NaiveTime (value encoding deserialization)
                        time:   [3.5495 ns 3.5583 ns 3.5675 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

bench Utf8 (len = 10) (key encoding deserialization)
                        time:   [37.607 ns 37.725 ns 37.841 ns]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

bench Utf8 (len = 10) (value encoding deserialization)
                        time:   [29.185 ns 29.248 ns 29.312 ns]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

bench Utf8 (len = 1000) (key encoding deserialization)
                        time:   [420.20 ns 421.55 ns 423.02 ns]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

bench Utf8 (len = 1000) (value encoding deserialization)
                        time:   [80.090 ns 80.290 ns 80.497 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

bench Utf8 (len = 10000) (key encoding deserialization)
                        time:   [17.046 µs 17.099 µs 17.153 µs]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

bench Utf8 (len = 10000) (value encoding deserialization)
                        time:   [4.8609 µs 4.8743 µs 4.8883 µs]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

bench Struct of Bool (len = 100) (key encoding deserialization)
                        time:   [2.2646 µs 2.2700 µs 2.2756 µs]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

bench Struct of Bool (len = 100) (value encoding deserialization)
                        time:   [2.1774 µs 2.1841 µs 2.1914 µs]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

bench List of Bool (len = 100) (key encoding deserialization)
                        time:   [1.2371 µs 1.2404 µs 1.2437 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

bench List of Bool (len = 100) (value encoding deserialization)
                        time:   [2.1033 µs 2.1079 µs 2.1129 µs]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

File: src/common/src/util/value_encoding/mod.rs [hunk 1]
```diff
@@ -89,50 +97,63 @@ fn serialize_value(value: ScalarRefImpl<'_>, mut buf: impl BufMut) {
...
     buf.put_u32_le(nano);
 }
 
-fn serialize_decimal(decimal: &Decimal, mut buf: impl BufMut) {
+fn serialize_decimal(decimal: &Decimal, buf: &mut impl BufMut) {
     buf.put_slice(&decimal.unordered_serialize());
 }
 
-fn deserialize_value(ty: &DataType, mut data: impl Buf) -> Result<Datum> {
+fn deserialize_value(ty: &DataType, data: &mut impl Buf) -> Result<Datum> {
```
Comment (BugenZhao) [inline src/common/src/util/value_encoding/mod.rs:156]: I guess you can pass a mutable reference of slice/cursor as `impl Buf`. `&mut impl Buf` sounds weird. 🤔

Comment (BugenZhao) [reply src/common/src/util/value_encoding/mod.rs:156]: https://docs.rs/bytes/latest/bytes/trait.Buf.html#impl-Buf-for-%26mut%20T

Comment (waruto210) [reply src/common/src/util/value_encoding/mod.rs:156]: I also find it strange to use `&mut impl Buf`, but using reference of slice/cursor needs to change the signature of `serialize/deserialize_datum`.

File: src/common/src/util/value_encoding/mod.rs [hunk 2]
```diff
@@ -147,69 +168,72 @@ fn deserialize_value(ty: &DataType, mut data: impl Buf) -> Result<Datum> {
...
-    let mut bytes = vec![0; len as usize];
-    data.copy_to_slice(&mut bytes);
-    ScalarImpl::from_proto_bytes(&bytes, &data_type.to_protobuf())
-        .map_err(ValueEncodingError::InvalidStructEncoding)
+fn deserialize_struct(struct_def: &StructType, data: &mut impl Buf) -> Result<ScalarImpl> {
+    let field_values = struct_def
+        .fields
+        .iter()
+        .map(|field_type| inner_deserialize_datum(data, field_type))
+        .collect::<Result<Vec<Datum>>>()?;
```
Comment (BugenZhao) [inline src/common/src/util/value_encoding/mod.rs:183]: For example, I think this should work.

```suggestion
fn deserialize_struct(struct_def: &StructType, data: impl Buf) -> Result<ScalarImpl> {
    let field_values = struct_def
        .fields
        .iter()
        .map(|field_type| deserialize_datum(&mut data, field_type))
        .collect::<Result<Vec<Datum>>>()?;
```

Comment (waruto210) [reply src/common/src/util/value_encoding/mod.rs:183]: In fact, that's what I did in the beginning. But this is a recursive invocation and the number of uses of `&mut buf` overflows.
569e6a660df963bea6a70ac7abc67fd6

**Comment by BugenZhao** [comment]
```
bench List of Bool (len = 100) (key encoding deserialization)
                        time:   [1.2371 µs 1.2404 µs 1.2437 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

bench List of Bool (len = 100) (value encoding deserialization)
                        time:   [2.1033 µs 2.1079 µs 2.1129 µs]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

This is intersting. Maybe it's caused by `collect`ing of value refs. How about this?

https://github.com/risingwavelabs/risingwave/blob/2dbbbade9fc4386db8a0c66a84f3b0accbdb87a4/src/common/src/array/list_array.rs#L493-L497

File: src/common/src/util/value_encoding/mod.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/util/value_encoding/mod.rs:156]: `&mut impl Buf` is also `impl Buf`. 😄 So there might be nothing to change.

**Comment by waruto210** [comment]
> ```
> bench List of Bool (len = 100) (key encoding deserialization)
>                         time:   [1.2371 µs 1.2404 µs 1.2437 µs]
> Found 1 outliers among 100 measurements (1.00%)
>   1 (1.00%) high mild
> 
> bench List of Bool (len = 100) (value encoding deserialization)
>                         time:   [2.1033 µs 2.1079 µs 2.1129 µs]
> Found 5 outliers among 100 measurements (5.00%)
>   4 (4.00%) high mild
>   1 (1.00%) high severe
> ```
> 
> This is intersting. Maybe it's caused by `collect`ing of value refs. How about this?
> 
> https://github.com/risingwavelabs/risingwave/blob/2dbbbade9fc4386db8a0c66a84f3b0accbdb87a4/src/common/src/array/list_array.rs#L493-L497

I'll try it locally first.

File: src/common/src/util/value_encoding/mod.rs (hunk 2 above)
Comment (BugenZhao) [reply src/common/src/util/value_encoding/mod.rs:183]: Okay... that makes sense. As the recursion depth of `Struct` scalar is unlimited, then we'll generate unlimited types with this function.

Comment (waruto210) [reply src/common/src/util/value_encoding/mod.rs:183]: Here is `deserialization`, and `iter_elems_ref!` is used for serialization.

**Comment by waruto210** [comment]
> 



> ```
> bench List of Bool (len = 100) (key encoding deserialization)
>                         time:   [1.2371 µs 1.2404 µs 1.2437 µs]
> Found 1 outliers among 100 measurements (1.00%)
>   1 (1.00%) high mild
> 
> bench List of Bool (len = 100) (value encoding deserialization)
>                         time:   [2.1033 µs 2.1079 µs 2.1129 µs]
> Found 5 outliers among 100 measurements (5.00%)
>   4 (4.00%) high mild
>   1 (1.00%) high severe
> ```
> 
> This is intersting. Maybe it's caused by `collect`ing of value refs. How about this?
> 
> https://github.com/risingwavelabs/risingwave/blob/2dbbbade9fc4386db8a0c66a84f3b0accbdb87a4/src/common/src/array/list_array.rs#L493-L497

Here is `deserialization`, and `iter_elems_ref!` is used for serialization.

**Comment by BugenZhao** [comment]
I'm drunk 🍻. But it will be nice to use `iter_elems_ref` for serializing list & struct.

**Review by BugenZhao — APPROVED** [review]

**Comment by waruto210** [comment]
> I'm drunk 🍻. But it will be nice to use `iter_elems_ref` for serializing list & struct.

but it's hard to get the length of the list before collecting.

**Pull request merged**